copilot: Fix config dir logic to support Flatpak environments

imumesh18 opened the pull request:
Closes #30784

In github copilot we were not handling the config path correctly for FLATPAK.

* Only tested on mac don't have access to other platform. But this should work on other platform as well. It follows the similar pattern seen in zed config path resolution.
- [x] Macos
- [ ] Linux
- [ ] Linux with Flatpak
- [ ] Windows

Release Notes:

- Fix copilot config detection for flatpack

imumesh18 committed:
copilot: Update config dir logic to support Flatpak environments

vixalien commented on crates/copilot/src/copilot_chat.rs:
```diff
@@ -424,12 +423,14 @@ pub fn copilot_chat_config_dir() -> &'static PathBuf {
-        if cfg!(target_os = "windows") {
-            home_dir().join("AppData").join("Local")
+        let config_dir = if cfg!(any(target_os = "linux", target_os = "freebsd")) {
+            std::env::var("FLATPAK_XDG_CONFIG_HOME")
```
I think you need to use `XDG_CONFIG_HOME`.

```
flatpak run --command=sh dev.zed.Zed
[📦 dev.zed.Zed ~]$ echo $XDG_CONFIG_HOME
/home/mona/.var/app/dev.zed.Zed/config
```

imumesh18 replied:
It should fallback to that if this env is not set. dir::config_path takes care of it.

imumesh18 replied:
Got it. I just made sure it's in inline with zed config: https://github.com/zed-industries/zed/blob/main/crates/paths/src/paths.rs#L75 resolution. Not sure why it exists or reason behind it for case of zed. But yes eventually it fallbacks to $XDG_CONFIG_HOME(which is handled by `dirs::config_dir()`) if it doesn't find FLATPAK_XDG_CONFIG_HOME env. So for now it should safely resolve and work. Will wait for some from zed to weigh in on whether or not we need.

mkungla replied:
@vixalien this was introduced way before I revamped  paths see e.g. https://github.com/zed-industries/zed/commit/81475ac4cde2d7481d620ef22e1d9c3148b38ec2

imumesh18 commented:
@vixalien builds are disabled by default someone from zed team needs to add run bundling tag to enable builds.

vixalien commented:
okie cool I'll test it when a build happens...

imumesh18 commented:
Hey @vixalien in case you want to test it you can also try building this on local. Here is the flatpack build instructions: https://zed.dev/docs/development/linux#flatpak

bennetbo commented:
Hey, gonna start a build now so that people can test

vixalien commented:
thanks !

On Wed, 18 Jun 2025 at 22:24, Bennet Bo Fenner ***@***.***>
wrote:

> *bennetbo* left a comment (zed-industries/zed#32901)
> <https://github.com/zed-industries/zed/pull/32901#issuecomment-2985590700>
>
> Hey, gonna start a build now so that people can test
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/zed-industries/zed/pull/32901#issuecomment-2985590700>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJB5FCOK2MJF4KKKPMYG5BL3EHDIPAVCNFSM6AAAAAB7REYOBGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDSOBVGU4TANZQGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

keramblock commented:
Build failed on some kind of upload

bennetbo commented:
Linux x86_64: https://github.com/zed-industries/zed/actions/runs/15743017685/artifacts/3358369992 
Linux arm64: https://github.com/zed-industries/zed/actions/runs/15743017685/artifacts/3358452897

vixalien commented:
Just built a flatpak with the build artifact, and sadly it does not work:

![image](https://github.com/user-attachments/assets/305924ed-0e30-496d-9d88-fb60d51a9e44)

![image](https://github.com/user-attachments/assets/9a5e9101-b915-4f3b-ab3a-6f7b2a5ecddc)

```
Zed: v0.193.0 (Zed Dev 80cf9ca8d208d3d6650ffdf4ec28ce4b64e93dc0) 
OS: Linux Wayland org.freedesktop.platform 24.08
Memory: 31.1 GiB
Architecture: x86_64
GPU: Intel(R) Iris(R) Xe Graphics (TGL GT2) || Intel open-source Mesa driver || Mesa 25.1.3 (git-ba95e694fe)
```

vixalien commented:
Also, it's a bit sad that Zed's CI doesn't automatically generate a flatpak, since it just needs to repackage the build artefacts from x86 and arm.

vixalien commented:
Okay there's a 30s~ delay.

Maybe Zed can improve the UX a bit by showing a spinner somewhere

imumesh18 committed:
Update config dir logic to use XDG_CONFIG_HOME on Unix

Switch to LocalAppData on Windows and remove Flatpak-specific logic. Now prefer
XDG_CONFIG_HOME if set, otherwise fallback to ~/.config.

imumesh18 commented:
Hey @bennetbo I have updated the logic a bit more and simplified it by checking the lsp source code how it looks for config path to store github-copilot. Below is the details. Let me know what do you think about current approach.

| Platform | Condition | Directory Path | Example |
|----------|-----------|----------------|---------|
| **Windows** | Always | `%LOCALAPPDATA%\github-copilot` | `C:\Users\username\AppData\Local\github-copilot` |
| **Linux/FreeBSD** | `XDG_CONFIG_HOME` set | `${XDG_CONFIG_HOME}/github-copilot` | `/home/username/.local/config/github-copilot` |
| **Linux/FreeBSD** | `XDG_CONFIG_HOME` not set | `~/.config/github-copilot` | `/home/username/.config/github-copilot` |
| **macOS** | `XDG_CONFIG_HOME` set | `${XDG_CONFIG_HOME}/github-copilot` | `/Users/username/.local/config/github-copilot` |
| **macOS** | `XDG_CONFIG_HOME` not set | `~/.config/github-copilot` | `/Users/username/.config/github-copilot` |
| **Other Unix-like** | `XDG_CONFIG_HOME` set | `${XDG_CONFIG_HOME}/github-copilot` | `/home/username/.local/config/github-copilot` |
| **Other Unix-like** | `XDG_CONFIG_HOME` not set | `~/.config/github-copilot` | `/home/username/.config/github-copilot` |

bennetbo reviewed (approved):
Thanks for investigating this 😄 Great work! Also thanks for testing to everyone involved

imumesh18 commented:
@bennetbo can you cherry pick this to preview please for next patch fix of preview.

bennetbo commented:
/cherry-pick v0.192.x

bennetbo commented:
/cherry-pick v0.191.x
